arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+    if (idx != id) {
+      auto& other = pool._threadStates[idx];
+      if (other->pushQueue.load(std::memory_order_relaxed) != nullptr) {
+        auto* item =
+            other->pushQueue.exchange(nullptr, std::memory_order_acquire);
+        if (item) {
+          std::unique_lock guard(mutex);
+          auto res = pushMany(item);
+          // remember this thread so next time we try to steal from it first
+          stealHint.store(idx, std::memory_order_relaxed);
+          return res;
+        }
+      }
+
+      if (other->queueSize.load(std::memory_order_relaxed) > 0) {
+        // stealing is only best effort - try to acquire lock, but don't bother
+        // if we don't get it right away
+        std::unique_lock guard(other->mutex, std::try_to_lock);
+        if (guard.owns_lock() && !other->_queue.empty()) {
+          auto toSteal = std::max<std::size_t>(1, other->_queue.size() / 2);
+
+          auto work = std::move(other->_queue.back());
+          TRI_ASSERT(work != nullptr);
+          other->_queue.pop_back();
+          if (toSteal > 1) {
+            std::unique_lock guard2(mutex);
+            for (std::size_t i = 1; i < toSteal; ++i) {
+              _queue.push_front(std::move(work));
+
+              TRI_ASSERT(other->_queue.size() > 0);
+              work = std::move(other->_queue.back());
+              TRI_ASSERT(work != nullptr);
+              other->_queue.pop_back();
+            }
+            queueSize.store(_queue.size(), std::memory_order_relaxed);
+          }
+          other->queueSize.store(other->_queue.size(),
+                                 std::memory_order_relaxed);
+
+          // remember this thread so next time we try to steal from it first
+          stealHint.store(idx, std::memory_order_relaxed);
+          return work;
+        }
+      }
+    }
+    if (this->pushQueue.load(std::memory_order_relaxed) != nullptr) {
+      // work has been pushed into our queue -> no need to still anything
+      return {};
+    }
+    idx += stepSize;
+    if (idx >= pool.numThreads) {
+      idx -= pool.numThreads;
+    }
+  }
+  // nothing found - reset the hint so next time we start at a random index
+  stealHint.store(NoHint, std::memory_order_relaxed);
+  return {};
+}
+
+WorkStealingThreadPool::WorkStealingThreadPool(const char* name,
+                                               std::size_t threadCount,
+                                               ThreadPoolMetrics metrics)
+    : numThreads(threadCount),
+      _metrics(metrics),
+      _threads(),
+      _threadStates(threadCount),
+      _latch(threadCount) {
+  auto coprimes = calculateCoprimes(threadCount);
+  std::size_t coprimeIdx = 0;
+  for (std::size_t i = 0; i < threadCount; ++i) {
+    _threads.emplace_back(std::jthread([this, i, &coprimes, coprimeIdx]() {
+      // we intentionally create the ThreadStates _inside_ the thread
+      _threadStates[i] =
+          std::make_unique<ThreadState>(i, coprimes[coprimeIdx], *this);
```

Comment:
nit: the separate `coprimeIdx` variable is not needed and could be substituted by just `i % coprimes.size()`.